Add currency units to the Number with Units interaction

vibhor98 opened the pull request:
This PR adds currency units (base + sub units) to the Number with Units interaction.
* `Dollar` as base unit and `cent` as its sub-unit.
* `Rupees` as base unit and `paisa` as its sub-unit.

<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include 
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue 
  - when this PR is merged.
  -->

## Checklist
- [ ] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes.
- [ ] The PR explanation includes the words "Fixes #bugnum: ...".
- [x] The linter/Karma presubmit checks have passed.
  - These should run automatically, but if not, you can manually trigger them locally using `python scripts/pre_commit_linter.py` and `bash scripts/run_frontend_tests.sh`.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR follows the [style guide](https://github.com/oppia/oppia/wiki/Coding-style-guide).
- [x] The PR is assigned to an appropriate reviewer.
  - If you're a new contributor, please ask on [Gitter](https://gitter.im/oppia/oppia-chat) for someone to assign a reviewer.
  - If you're not sure who the appropriate reviewer is, please assign to the issue's "owner" -- see the "talk-to" label on the issue.

vibhor98 committed:
Add currency units.

vibhor98 committed:
Fix failing frontend tests.

kevinlee12 commented:
Hi @vibhor98, some initial thoughts:
- Do add currency to the "Possible units formats" table.
- The following is a bit bothersome, since it didn't match what I inputted ($0). I would not mind if the spaces were changed, but the notion that results is quite strange:
![image](https://user-images.githubusercontent.com/5758256/42196540-5c2d8a8c-7e33-11e8-9e95-321e9bff4a88.png)
- The point is driven home in the learner view where students may be even more confused and definitely may not know what are exponents: 
![image](https://user-images.githubusercontent.com/5758256/42196593-a9f2d77c-7e33-11e8-9389-3f06cac4b2e3.png)
- Cents doesn't register properly, so you do want to look at that. 

That being said, I do like the fact that it validates as I'm typing. As for the second and third points, you might want to store what the user has inputted (if need be, evaluate on the fly), so you can avoid user confusion.

vibhor98 commented:
Hi @kevinlee12, the problem in 2nd and 3rd point is because `math.js` doesn't support special characters in the custom units (contrary to the case of currency units). Also, it doesn't allow writing unit and then value like `$ 1`. So, I had to change `$` to `dollar` first and that's why it appeared strangely. To store a copy what the user has inputted can resolve this problem. Also, cents appear to be working well.

kevinlee12 commented:
Hmm, definitely get what you are doing there, but I don't think it'll fly with our users. can you think of a way where you override what is shown when the user inputs a currency?

vibhor98 commented:
Definitely, I can remove the `^1` unnecessary sign. Then, it will be `1 dollar` for `$ 1`. Will it work? I know it is still not exactly what the learner has inputted.

kevinlee12 commented:
Honestly still hesitant about it, you'll have to run some user trials to see if this is an acceptable answer. Ideally, you should store what the user has and just override it in math.js. However, I'll defer that technical discussion to @prasanna08.

vibhor98 commented:
@prasanna08, what are your suggestions?

seanlip commented:
In general, what is shown to the learner needs to visually match what the learner types.

prasanna08 commented:
@vibhor98 can't we replace `$` with `dollar` (just like `per` and `/`) for our internal representation (and append it at back rather than the front)? This creates two different strings, one which is shown to learner and other that we use for our own representation (i.e. math.js).

prasanna08 commented on core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.js:
```diff
@@ -279,12 +285,37 @@ oppia.factory('UnitsObjectFactory', [function() {
+  Units.createCurrencyUnits = function() {
+    // Creates user-defined currency (base + sub) units.
+    math.createUnit('dollar', {aliases: [
+      'dollars', 'Dollar', 'USD', '$', 'Dollars']});
```
You said that special symbol such as `$` is not allowed but you are still using it as an alias. Does it work?

vibhor98 replied:
In aliasing, it is allowed. But when you'll actually type `$`, it won't be allowed even though it is in alias. May be it's a bug in `math.js` that it is allowed in alias.

prasanna08 replied:
Hmm, strange. What is the point of alias, then?

vibhor98 replied:
Yes, `$` is not allowed. But other aliases are working well.

prasanna08 replied:
Well, then, I suggest removing `$`. It leads to confusion for developers.

prasanna08 commented on core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.js:
```diff
@@ -279,12 +285,37 @@ oppia.factory('UnitsObjectFactory', [function() {
+    // Special symbols need to be replaced as math.js doesn't support custom
+    // units starting with special symbols. Also, it doesn't allow units
+    // followed by a number as in the case of currency units.
+    if (units.includes('$')) {
```
Maybe do this when you are checking equivalency using math.js otherwise there is no need to do this replacement.

vibhor98 replied:
Validation is also taking place before answer submission. We ned to replace it here as well.

prasanna08 replied:
Yep. You can create some function in UnitsObjectFactory which gives you math.js compatible representation format. So that could work, right?

vibhor98 replied:
Done.

prasanna08 commented on core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.js:
```diff
@@ -279,12 +285,37 @@ oppia.factory('UnitsObjectFactory', [function() {
+      units = units.replace('$', '');
+      units = 'dollar ' + units;
+    }
+    if (units.includes('Rs') || units.includes('₹')) {
```
Ditto.

prasanna08 replied:
Is `Rs` not supported by math.js? I am asking because it is not a special symbol.

vibhor98 replied:
Yes, it is supported. But `math.js` doesn't allow any unit followed by value. Like: `Rs 5` isn't allowed. So I've to change it as well.

prasanna08 replied:
I see. That is okay.

vibhor98 commented:
@prasanna08, This is exactly what I'm doing. The verification is going on smoothly. But what @kevinlee12 pointed out was- In the response, the learner will get the `1 dollar` even though he has entered `$ 1`.

prasanna08 commented:
> @prasanna08, This is exactly what I'm doing. The verification is going on smoothly. But what @kevinlee12 pointed out was- In the response, the learner will get the 1 dollar even though he has entered $ 1.

Yep. I just saw the code. Look at the comment.

vibhor98 committed:
Fix currency response.

vibhor98 commented:
PTAL at my latest commit. Now the response is same as the input given. Like response: `$ 1` for `$ 1` input. I've modified help table too and removed unnecessary addition of `^1` dimension to the units. 😄 
![screen shot 2018-07-03 at 8 43 12 pm](https://user-images.githubusercontent.com/26625980/42231856-539ac5f6-7f0a-11e8-8335-668140edb35d.png)

prasanna08 commented on core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.js:
```diff
@@ -65,6 +65,22 @@ oppia.factory('NumberWithUnitsObjectFactory', [
       return numberWithUnitsString;
     };
 
+    NumberWithUnits.prototype.toCompatibleString = function() {
```
`toMathjsCompatibleString`

vibhor98 replied:
Done

prasanna08 commented on core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.js:
```diff
@@ -279,14 +301,45 @@ oppia.factory('UnitsObjectFactory', [function() {
+    math.createUnit('paise', {definition: '0.01 rupees', aliases: ['paisa']});
+  };
+
+  Units.makeUnitsCompatible = function(units) {
```
`toMathjsCompatibleString`

vibhor98 replied:
Done

prasanna08 reviewed (changes requested):
I will give it a try later on the local server.

vibhor98 committed:
Made review changes.

prasanna08 commented:
Hi, @vibhor98 

I was testing interaction on local server and found one interesting issue.

The rules don't consider `$ 5` and `5 dollars` as equal units. This is because $ and dollar are different in exact (===) equality match and rules failing makes sense because we are checking for exact equality. But that should be relaxed for currency case because $ and dollar(s) should be seen as equal even if they are not exactly equal.

I suggest that you convert them to a singular representation before moving on to check its equality `($, dollar, dollars) ==> (dollars)` (same for Rs).

**EDIT**
NOTE: Conversion should happen only within equal checking part. It should not do any permanent changes to answer. So that we can show users what they entered.

prasanna08 commented on core/templates/dev/head/domain/objects/NumberWithUnitsObjectFactory.js:
```diff
@@ -65,6 +65,22 @@ oppia.factory('NumberWithUnitsObjectFactory', [
 
+    NumberWithUnits.prototype.toMathjsCompatibleString = function() {
+      var numberWithUnitsString = '';
+      var unitsString = UnitsObjectFactory.fromList(this.units).toString();
```
Consider calling `UnitsObjectFactory.fromList(this.units).toMathjsCompatibelString();` directly. Use toString function withing it to get unitString.

vibhor98 commented:
Hi @prasanna08, I understand your concern and knew this interesting issue beforehand. In fact I tried to remove it but faced the issue. Actually, in `Equal to` check, we're directly comparing both dicts that is coming from the backend. So, we can't actually change it before it is saved as dict. One possible solution I can think of is- Convert this `dicts` to `String`, then to `CompatibleString` and then, to `NumberWithUnits` object followed by `toDict` function. Is this much complexity fine or there can be other way too?

NOTE: We can't directly compare two strings in `Equal to` rule as order of units may not be necessarily  same. Eg- `N m` and `m N` are equal.

prasanna08 commented:
In that case, convert units to the compatible string first and then convert the compatible string back into unitlList. Then compare units and real/fraction separately. Would that work?

vibhor98 committed:
Made equality check flexible for currency units.

vibhor98 commented:
Done @prasanna08!  Now the check is flexible for currency units.
